Added a dockerfile for the build dependencies

nath1416 opened the pull request:
This Pr adds a dockerfile to encapsulate the python2 build dependencies of the build script. This allows the script to be run on any platforms that has docker installed and prevents installing python2 directly on your computer. 

To use the container to patch the firmware, you only need to build the container and then run it as showed in the readme.

Let me know what you think of it!

bmeares commented:
@nath1416 thanks for putting together this PR! I was trying to run this tool on Fedora and trying to get Python 2 to run was a nightmare. Perhaps I'll put together a PR or fork to rewrite some of the utilities with a more modern Python stack.

kendallgoto commented:
i have a copy written natively in Go that i'm (slowly) working to get cleaned up and replace this repo, fwiw.

thanks for the Dockerfile - will try to review + send some feedback this week.

nath1416 commented:
Thanks! 
Just flashed my dl160g9 today and it worked like a charm.

kendallgoto commented on Dockerfile:
```diff
@@ -0,0 +1,27 @@
+FROM ubuntu:22.04
+
+RUN apt-get update && \ 
```
can merge these apt-get steps

nath1416 replied:
I was actually able to remove the first line by switching to debian as the base image

kendallgoto replied:
could we just use debian:bullseye-slim here rather than 11.11?

nath1416 replied:
Yes good idea, changed it.

kendallgoto commented on Dockerfile:
```diff
@@ -0,0 +1,27 @@
+FROM ubuntu:22.04
```
ubuntu:22.04 feels very heavy to use as a base image - i'd think even just alpine:3 would be okay here

nath1416 replied:
You're right, I tried alpine but got multiple dependency error when installing packages via pip.
I was able to use debian:11.11 instead which reduces the size of the container.

kendallgoto commented on README.md:
````diff
@@ -28,14 +28,27 @@ python2 -m virtualenv venv
 ```
+Or if you are using the docker container. Make sure you are in the `ilo4_unlock` directory first.
+```bash
+docker run --rm -it -v $(pwd):/app ilo4_unlock:latest init # download necessary HPE binaries
````
can drop repeated comments

kendallgoto reviewed (changes requested):
looks good; some nits to clean up on the Dockerfile.

I wonder if the default behavior on the dockerfile should just be running it + it will run the default patches (init + 277) without needing user input, rather than just running the `./build.sh` interface directly - to simplify experience, and also possibly eliminate the need for the volume mount which complicates the setup for those less familiar w/ docker

nath1416 commented:
I changed the default behavior, when the container is started without arguments, it will now run the build script with init + latest if no arguments are passed.
I changed the entrypoint file for this  behavior.

But I don't know if it's possible to remove the volume, the patch firmware needs to accessible somehow to the user...

Let me know what you think!

0n1cOn3 commented on Dockerfile:
From https://github.com/kendallgoto/ilo4_unlock/pull/36#issuecomment-2869061835, about docker volumes and mountpoints:

You can declare `/app/build` as volume in dockerfile like:

`VOLUME /app/build`

And add later the command for docker and/or docker compose like:

```bash
docker run -d \
  --name ilo4_unlock_builder \
  -v /mnt/ilo4_unlock_builder:/app/build \
  ilo4_unlock_builder:latest
```

```yaml
version: "3.8"

services:
  ilo4_unlock_builder:
    image: ilo4_unlock_builder:latest
    container_name: ilo4_unlock_builder
    volumes:
      - /mnt/il4_unlock_build:/app/build
    restart: unless-stopped
```

kendallgoto replied:
the VOLUME directive is only really useful if you're using named volumes and not for host-bindings afaik. in this usage, it's actually not getting used and is getting overwritten by the cli argument. my intention was to eliminate the cli argument entirely just to reduce the complexity of the command.

In this case, I had actually intended using docker build --out and eliminating the run entirely; but I think doing that in a single container that also allows the CLI is cumbersome. So I think this is OK with me.

kendallgoto reviewed (approved):
reads good to me; one little thing i commented on. otherwise ill test it out real quick and merge shortly

0n1cOn3 commented:
Ahh, I see now where it is going to.. 

That could mean that the snippet would do the job instead of my previous comment:

```dockerfile
FROM scratch AS output
COPY --from=builder /app/ilo4_unlock/build /ilo4_unlock
```

nath1416 commented:
Just address your comment, I also needed to add this to the Dockerfile because I got strange error from git.

`RUN git config --global --add safe.directory /app`
